### ISSUE TITLE
[QUESTION] Predicate Order matter ?

### ISSUE THREAD

**Issue by toumir** [litedb-org/LiteDB#1829]
I wonder if the order is matter  when querying the database ?
The **results** of the following two queries are **different**, can someone explain that to me,

LiteDB version:  **master branch**

```
void Main()
{
	var cs = new ConnectionString
	{
		Filename = @"c:\mydb.db",
		Password = @"MyP@55w0rd"
	};

	using (var db = new LiteDatabase(cs))
	{
		var col = db.GetCollection<Product>("Prods");
		col.DeleteAll();
		
		#region
		col.Insert(new Product { Id = 4, Name = "Vantage" });
		col.Insert(new Product { Id = 5, Name = "Corvette" });
		col.Insert(new Product { Id = 7, Name = "Caprice" });
		col.Insert(new Product { Id = 6, Name = "3000GT" });
		#endregion
		
		col.EnsureIndex(c => c.Name);
		
		Expression<Func<Product, bool>> p1 = p=>false; 
		Expression<Func<Product, bool>> p2 = p=>p.Name.Contains("a");
		
		//1 query
		col
		.Query()
		.Where(p2.Or(p1))
		.Count();//<== cout == 2
		
		//2 query
		col
		.Query()
		.Where(p1.Or(p2))
		.Count();//<== cout == 0
	}
}
// Define other methods, classes and namespaces here
public class Product
{
	public int Id{get;set;}
	public string Name{get;set;}
}

```

Thank you in advance

_Originally posted by @toumir in https://github.com/mbdavid/LiteDB/issues/1824#issuecomment-697361182_

**Comment by lbnascimento** [comment]
@toumir This will be a long explanation.

LiteDB implements its own `ExpressionVisitor`. Basically, what it does is transform a regular Linq expression into a `BsonExpression`, which is a wrapper for an expression that can be used with a mapped `BsonDocument`. When parsing the original expression, it needs to know whether a certain parameter refers to the root document (`$`) or to another scope (`@`). Up to v5.0.7, this comparison was done by comparing the parameter name. This approach would worked fine for your situation, but it didn't work for expressions that had parameters with the same name in different scopes (e.g.: `x => x.Names.Any(x => x.Length > 5)`).

In v5.0.8, it was changed so that it stores and compares the actual `ParameterExpression`. And given that your expressions `p1` and `p2` were created separately, they technically have different parameters. In other words, the parameter `p` in `p2` and the parameter `p` in `p2` are different objects in memory, and our parser doesn't have any way to know that they should be treated as the same parameter.

I tried to get this "merged" expression to work with LiteDB. Replacing the left-side `p` is simple, but you'd have to replace all references to `p` in the right side too, which is much harder. It is definitely possible, but it would require a lot of boilerplate code.

I'm honestly not sure what to do. One possibility is to keep everything as it is and document this limitation. Another possibility is to revert to the name comparison, which would break compatibility for some people, but it at least has a simpler workaround. I'm not sure there's a simple solution that would get both cases working at the same time.

**Comment by toumir** [comment]
@lbnascimento thank you for following and for the deep explanation,

**"I'm honestly not sure what to do."**
what I'm doing is complicated, but I'll explain and simplify as much as possible.

In my mobile application, I have more than 11 pages (and the number will increase in the future) we call it **Listing**, each Listing has its own filters, a page may or may not have the same filters,

each listing has two types of filters, one for the **keyword** filter and the other for the **Extra filters**
let search by keyword and use the Extra filters,

before calling the search API, I must to combine the two predicates, 

to display the number of elements found I have to do this:
```
var count = col
.Query ()
.Where (Predicate1.And (Predicate2))
.Count ();
```
in my case the count is always 0

to sum up
"`.Where (Predicate1.Or (Predicate2)).Count()`" and "`.Where (Predicate2.Or (Predicate1)).Count()`" 
do not give the same results
thank you in advance
![filters](https://user-images.githubusercontent.com/10366325/94177574-a7f84480-fe91-11ea-9479-15991017b058.png)

**Comment by toumir** [comment]
Hi,
@lbnascimento  explained well how LiteDB works under the hood, 
Since I made a big refactor and all my queries are based on predicate, I am **stuck** for more than 3 weeks, can someone help me find a solution to this situation, 
Did someone manage how to combine two predicate and use it in LiteDB ?
thanks developers

**Comment by toumir** [comment]
this workaround worked for me 
```
 public IList<Expression<Func<Product, bool>>> GetListPredicates()
{
var lstExpr = new List<Expression<Func<Product,bool>>> {p => p.IsChrgDechrg};
if(condition)
{
	lstExpr.Add(p => p.IsSynch);
}
if(condition2)
{
	lstExpr.Add(p => p.Name.Constains("a"));
}
....
 return lstExpr;
}
```

**usage**

```
var queryable = Context.GetCollection<T>().Query();
foreach (var p in GetListPredicates())
{
    queryable.Where(p);
 });			
queryable.ToList();
```

**Comment by AlBannaTechno** [comment]
> 
> 
> @toumir This will be a long explanation.
> 
> LiteDB implements its own `ExpressionVisitor`. Basically, what it does is transform a regular Linq expression into a `BsonExpression`, which is a wrapper for an expression that can be used with a mapped `BsonDocument`. When parsing the original expression, it needs to know whether a certain parameter refers to the root document (`$`) or to another scope (`@`). Up to v5.0.7, this comparison was done by comparing the parameter name. This approach would work fine for your situation, but it doesn't work for expressions that have parameters with the same name in different scopes (e.g.: `x => x.Names.Any(x => x.Length > 5)`).
> 
> In v5.0.8, it was changed so that it stores and compares the actual `ParameterExpression`. And given that your expressions `p1` and `p2` were created separately, they technically have different parameters. In other words, the parameter `p` in `p2` and the parameter `p` in `p2` are different objects in memory, and our parser doesn't have any way to know that they should be treated as the same parameter.
> 
> I tried to get this "merged" expression to work with LiteDB. Replacing the left-side `p` is simple, but you'd have to replace all references to `p` in the right side too, which is much harder. It is definitely possible, but it would require a lot of boilerplate code.
> 
> I'm honestly not sure what to do. One possibility is to keep everything as it is and document this limitation. Another possibility is to revert to the name comparison, which would break compatibility for some people, but it at least has a simpler workaround. I'm not sure there's a simple solution that would get both cases working at the same time.


@lbnascimento thank you for your explaination

I work with similar issue but less complex compare to this, in `ReactiveUI.Validation` Packaage `My Own Fork`

so can you please refere me to this part in the source code `file/line-s`, i will try to solve it.

**Comment by toumir** [comment]
@AlBannaTechno maybe @lbnascimento answer you :)

**Comment by AlBannaTechno** [comment]
> 
> 
> this workaround worked for me
> 
> ```
>  public IList<Expression<Func<Product, bool>>> GetListPredicates()
> {
> var lstExpr = new List<Expression<Func<Product,bool>>> {p => p.IsChrgDechrg};
> if(condition)
> {
> 	lstExpr.Add(p => p.IsSynch);
> }
> if(condition2)
> {
> 	lstExpr.Add(p => p.Name.Constains("a"));
> }
> ....
>  return lstExpr;
> }
> ```
> 
> **usage**
> 
> ```
> var queryable = Context.GetCollection<T>().Query();
> foreach (var p in GetListPredicates())
> {
>     queryable.Where(p);
>  });			
> queryable.ToList();
> ```

@toumir

I Used similar solution for a long time and its is work perfectly,
But we still need to fix the bug in the library.

So i wait, @lbnascimento, and i will try to solve it.
